### PR TITLE
storage: add conformance tests and Kokoro CI

### PR DESCRIPTION
**Pull request by frankyn**
- [x] Add Storage Conformance tests for V4 Signed URLs.
- [x] Set up Kokoro CI on the repository.
- [x] Tests pass.

### CODE REVIEW

File: storage/v1/conformance_tests.json [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    {
+      "description": "Simple headers",
+      "bucket": "test-bucket",
+      "object": "test-object",
+      "method": "GET",
+      "expiration": "10",
+      "timestamp": "2019-02-01T09:00:00Z",
+      "expectedUrl": "https://storage.googleapis.com/test-bucket/test-object?X-Goog-Algorithm=GOOG4-RSA-SHA256\u0026X-Goog-Credential=test-iam-credentials%40dummy-project-id.iam.gserviceaccount.com%2F20190201%2Fauto%2Fstorage%2Fgoog4_request\u0026X-Goog-Date=20190201T090000Z\u0026X-Goog-Expires=10\u0026X-Goog-SignedHeaders=bar%3Bfoo%3Bhost\u0026X-Goog-Signature=68ecd3b008328ed30d91e2fe37444ed7b9b03f28ed4424555b5161980531ef87db1c3a5bc0265aad5640af30f96014c94fb2dba7479c41bfe1c020eb90c0c6d387d4dd09d4a5df8b60ea50eb6b01cdd786a1e37020f5f95eb8f9b6cd3f65a1f8a8a65c9fcb61ea662959efd9cd73b683f8d8804ef4d6d9b2852419b013368842731359d7f9e6d1139032ceca75d5e67cee5fd0192ea2125e5f2955d38d3d50cf116f3a52e6a62de77f6207f5b95aaa1d7d0f8a46de89ea72e7ea30f21286318d7eba0142232b0deb3a1dc9e1e812a981c66b5ffda3c6b01a8a9d113155792309fd53a3acfd054ca7776e8eec28c26480cd1e3c812f67f91d14217f39a606669d",
+      "headers": {
+        "BAR": "BAR-value",
```
Comment (jadekler) [inline storage/v1/conformance_tests.json]: Why are these included? I thought only x-goog are allowed. Our regex is:

```
canonicalHeaderRegexp    = regexp.MustCompile(`(?i)^(x-goog-[^:]+):(.*)?$`)
```

Everything else gets stripped out in our impl.

(it could be that our impl is wrong, though)

File: storage/v1/validator.go [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+	storage_v1_tests "./generated/storage/v1"
+)
+
+func main() {
+	args := os.Args
+	dir := args[1]
+
+	files, err := ioutil.ReadDir(dir)
+	if err != nil {
+		panic(err)
```
[the remark this answers is not in the document]
Comment (frankyn) [reply storage/v1/validator.go]: Acking, thanks!

File: storage/v1/validator.go [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+	}
+
+	for _, f := range files {
+		if !strings.Contains(f.Name(), "tests.json") {
```
Comment (jskeet) [inline storage/v1/validator.go]: I suspect we *may* want to go the other way, and assume every JSON file in the directory is a test unless it has some special name. (For example, we could rename the service account to "test_service_account.not-a-test.json".

This isn't a blocker though. I'm keen to get the PR merged and iterate later :)

File: validator.sh [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+# Prerequisites:
+# 1. golang compiler must be installed.
+# 2. protoc must be installed.
+# 3. protoc-gen-go must be installed.
+
+# Generate Go files in a tmp directory so they're automatically git-ignored.
+# They'll still be compiled.
+rm -rf storage/v1/generated
+mkdir -p storage/v1/generated
+protoc --go_out=storage/v1/generated storage/v1/test.proto
```
Comment (jskeet) [inline validator.sh]: Do we not need to add an include path for the built-in protos? (For timestamp.) If it works, I'm happy...

File: storage/v1/validator.go (hunk 3 above)
Comment (jadekler) [reply storage/v1/validator.go]: +1, that seems reasonable. We could also add a simple exclusion list in code, since the list of .json files that aren't tests will be small (possibly just the service account?).

File: validator.sh (hunk 4 above)
Comment (jadekler) [reply validator.sh]: I was able to generate it without the import, but now that you mention it that is a bit odd. Maybe it's baked into the protoc golang plugin somehow...? o.O

Comment (frankyn) [reply validator.sh]: I'm not sure either. The generated proto Go code contains the necessary Timestamp import.

File: storage/v1/validator.go (hunk 3 above)
Comment (frankyn) [reply storage/v1/validator.go]: Updated.

File: storage/v1/conformance_tests.json (hunk 1 above)
Comment (frankyn) [reply storage/v1/conformance_tests.json]: For v4 you can define non x-goog headers such as Content-Type. They would populate signedHeaders.

Comment (jadekler) [reply storage/v1/conformance_tests.json]: Sorry, I don't grok. Content-Type is a string, not a map. How would I supply them as Content-Type?

Comment (frankyn) [reply storage/v1/conformance_tests.json]: I meant `Content-Type` as a generalization. Right now the Go surface provides a special parameter which meets v2 expectations given v2 doesn't use non x-goog headers in the signature. 

For v4, all headers are treated equally and signed for example Content-Type and x-goog headers.
Even though v4 handles headers differently the Go surface API didn't change (other languages did the same). 

If it helps we can walk through this more in a short meeting.

**Comment by frankyn** [comment]
Pending internal change before merging.. Thank you for your patience!

**Comment by frankyn** [comment]
Internal change was merged and Kokoro CI is enabled. @jadekler can I get approval or are you pending verifying the Go implementation?

**Comment by jadekler** [comment]
I've got the header sanitization working as these tests expect, but I'm getting a different signature for `Headers_should_be_trimmed`:

```
    --- PASS: TestSigningV4Conformance/Simple_GET (0.00s)
    --- PASS: TestSigningV4Conformance/Simple_PUT (0.00s)
    --- PASS: TestSigningV4Conformance/POST_for_resumable_uploads (0.00s)
    --- PASS: TestSigningV4Conformance/Vary_expiration_and_timestamp (0.00s)
    --- PASS: TestSigningV4Conformance/Vary_bucket_and_object (0.00s)
    --- PASS: TestSigningV4Conformance/Simple_headers (0.00s)
    --- FAIL: TestSigningV4Conformance/Headers_should_be_trimmed (0.00s)
        conformance_test.go:81: 
            want:	https://storage.googleapis.com/test-bucket/test-object?X-Goog-Algorithm=GOOG4-RSA-SHA256&X-Goog-Credential=test-iam-credentials%40dummy-project-id.iam.gserviceaccount.com%2F20190201%2Fauto%2Fstorage%2Fgoog4_request&X-Goog-Date=20190201T090000Z&X-Goog-Expires=10&X-Goog-Signature=1839511d6238d9ac2bbcbba8b23515b3757db35dfa7b8f9bc4b8b4aa270224df747c812526f1a3bcf294d67ed84cd14e074c36bc090e0a542782934a7c925af4a5ea68123e97533704ce8b08ccdf5fe6b412f89c9fc4de243e29abdb098382c5672188ee3f6fef7131413e252c78e7a35658825ad842a50609e9cc463731e17284ff7a14824c989f87cef22fb99dfec20cfeed69d8b3a08f00b43b8284eecd535e50e982b05cd74c5750cd5f986cfc21a2a05f7f3ab7fc31bd684ed1b823b64d29281e923fc6580c49005552ca19c253de087d9d2df881144e44eda40965cfdb4889bf3a35553c9809f4ed20b8355be481b92b9618952b6a04f3017b36053e15&X-Goog-SignedHeaders=collapsed%3Bhost%3Bleading%3Btrailing
            got:	https://storage.googleapis.com/test-bucket/test-object?X-Goog-Algorithm=GOOG4-RSA-SHA256&X-Goog-Credential=test-iam-credentials%40dummy-project-id.iam.gserviceaccount.com%2F20190201%2Fauto%2Fstorage%2Fgoog4_request&X-Goog-Date=20190201T090000Z&X-Goog-Expires=10&X-Goog-Signature=203c9c85fbeaacc318dac8761833e341a3f52a905ef1b1b46e62c742f466812887cd28412984f0cf0e134bed489f35df16dda38c42eaa317012221ae21bc3f768473dc593c3ca7acc24c0558d3a8085ff663846eff33b05c3176a15d7ebcc382cc5e800fe34b09fd630669f1bd1985a5636691dad518ab74e525d87ed1f92574fa40219ab884a13b1c1a8692ef12545dd2f7930a97e2a0ef00c9f2840cb2ec91f4afb1a075a9fa3933a91e8c24ce71e52ba987cdc760bb81f52843b7c40cfa2ca22f45dba73e2d05c6cf894e7d6e8d3e48bc5fd76eeddbb8a43795b36d6884c1da27c0baf7f09d3e7ace70a5a12bf7808a8f2fcd9631931b0e4efb0d5f7b5481&X-Goog-SignedHeaders=collapsed%3Bhost%3Bleading%3Btrailing
```

All query params seem to be identical except the signature. Any idea what gives?

**Comment by jadekler** [comment]
cc @jskeet ^

**Comment by jskeet** [comment]
Hmm. Can't say off the top of my head - will check that the signature matches what I get with the C# implementation tomorrow. (It's possible that I messed up when copying the data.)

**Comment by frankyn** [comment]
@jadekler could you add intermediate step of the "string_to_sign"?

**Comment by jadekler** [comment]
> @jadekler could you add intermediate step of the "string_to_sign"?

```
GOOG4-RSA-SHA256
20190201T090000Z
20190201/auto/storage/goog4_request
59c1ac1a6ee7d773d5c4487ecc861d60b71c4871dd18fc7d8485fac09df1d296
```

**Comment by jadekler** [comment]
Oh, wait, sorry! I posted the wrong set. Here's the actual thing:

```
=== RUN   TestSigningV4Conformance/Headers_should_be_trimmed
GOOG4-RSA-SHA256
20190201T090000Z
20190201/auto/storage/goog4_request
90d8a5287bc7604d3516a1f0cc4c0cdb37f8472721dc15f976ee725051ef9225
```

**Comment by frankyn** [comment]
I was able to reproduce your signature in the string to sign based on the information you provided over chat

```
GET
/test-bucket/test-object
X-Goog-Algorithm=GOOG4-RSA-SHA256&X-Goog-Credential=test-iam-credentials%40dummy-project-id.iam.gserviceaccount.com%2F20190201%2Fauto%2Fstorage%2Fgoog4_request&X-Goog-Date=20190201T090000Z&X-Goog-Expires=10&X-Goog-SignedHeaders=collapsed%3Bhost%3Bleading%3Btrailing
collapsed:abc    def
host:storage.googleapis.com
leading:xyz
trailing:abc

collapsed;host;leading;trailing
UNSIGNED-PAYLOAD
```

The issue is with the collapsed header value `collapsed:abc    def` it should be `collapsed:abc def` instead.

Collapsing the inline whitespace gives me the correct intermediate hash of the canonical request being `2b49a17175c9a50df40b20cdb5c3f626d056a2182f7d38fda396b85bb31135a0`.

**Comment by jskeet** [comment]
And lo, the value of conformance tests is proved :)

**Comment by jadekler** [comment]
> And lo, the value of conformance tests is proved :)

Yes! Very well done making these. :)

PS: Should tabs also get trimmed? Maybe we should add a test for that?

**Comment by frankyn** [comment]
It's just spaces for now @jadekler. There's no clear definition of whitespace for v4 signatures. The backend will do more than just spaces, but libraries don't verify that.

**Comment by jadekler** [comment]
> It's just spaces for now @jadekler. There's no clear definition of whitespace for v4 signatures. The backend will do more than just spaces, but libraries don't verify that.
> 
> Thanks so much for putting these together @jskeet, it really helped put all the libraries on the same page.

SG. I removed it, but now I see you added a test that checks tabs? Should I add it back? o.O

**Comment by frankyn** [comment]
I commented on your implementation. Apologies, I remembered wrong per documentation https://cloud.google.com/storage/docs/access-control/signed-urls#about-headers
